### PR TITLE
[typescript] Migrate from API to MCP

### PR DESCRIPTION
**Pull request by sgr-stripe**
### Summary

**Version change:** 8.1.0 -> 9.0.0

**Breaking change:** Migrate `@stripe/agent-toolkit` from direct Stripe API calls to MCP (Model Context Protocol) architecture. The toolkit now connects to `mcp.stripe.com` to fetch and execute tools.

Key changes:
- **Async initialization required**: Use `await createStripeAgentToolkit(config)` instead of `new StripeAgentToolkit(config)`
- **New shared infrastructure**: `ToolkitCore` base class, `StripeMcpClient`, `StripeClient`, `AsyncInitializer`
- **All framework adapters updated**: ai-sdk, langchain, openai, and modelcontextprotocol now use MCP
- **Tool names now snake_case**: e.g., `createCustomer` → `create_customer`
- **`@modelcontextprotocol/sdk` promoted to direct dependency** (from peer dependency)
- **Cleanup support**: All toolkits now have `close()` for proper resource cleanup
- **Restricted key warning**: `sk_*` keys emit deprecation warning; `rk_*` preferred

Files removed:
- `src/shared/api.ts` (replaced by `stripe-client.ts` + `mcp-client.ts`)
- `src/ai-sdk/tool.ts`, `src/langchain/tool.ts` (inlined into toolkits)

New files:
- `src/shared/mcp-client.ts` - MCP SDK wrapper for `mcp.stripe.com`
- `src/shared/stripe-client.ts` - Unified client (MCP for tools, SDK for billing)
- `src/shared/toolkit-core.ts` - Base class for all framework toolkits
- `src/shared/async-initializer.ts` - Reusable async init with promise locking
- `src/shared/schema-utils.ts` - JSON Schema to Zod conversion
- `src/shared/constants.ts` - Shared version and URL constants
- `MIGRATION.md` - User-facing migration guide

### Motivation

Centralizes tool definitions and execution on `mcp.stripe.com`, enabling server-side updates without SDK releases.

### Test plan

- [x] All changes in PR are covered by tests
- [x] Failures and edge cases tested

Integration test results, running parallel published SDK vs changes: [link to results](https://git.corp.stripe.com/sgr/stripe-ai-parity-tests/blob/main/typescript/results/parity-test-run-all-2026-01-29T21-32-30-000Z.txt)

New test coverage:
- `async-initializer.test.ts` - concurrent init, retry after failure, reset
- `mcp-client.test.ts` - connect, disconnect, tool calls, context handling
- `schema-utils.test.ts` - JSON Schema to Zod conversion

### CODE REVIEW

File: tools/typescript/src/modelcontextprotocol/toolkit.ts [hunk 1]
```diff
@@ -16,43 +21,138 @@ class StripeAgentToolkit extends McpServer {
...
+    });
+
+    this._configuration = configuration;
+
+    // MCP client for connecting to mcp.stripe.com
+    this._mcpClient = new StripeMcpClient({
+      secretKey,
+      context: {
+        account: configuration.context?.account,
+        customer: configuration.context?.customer,
```
Comment (matv-stripe) [inline tools/typescript/src/modelcontextprotocol/toolkit.ts:32]: ah, this is kind of a monkey wrench, this will not work as well

File: tools/typescript/src/modelcontextprotocol/toolkit.ts [hunk 2]
```diff
@@ -16,43 +21,138 @@ class StripeAgentToolkit extends McpServer {
...
       },
     });
 
-    this._stripe = new StripeAPI(secretKey, configuration.context);
+    // Keep Stripe SDK for registerPaidTool billing operations
+    this._stripe = new Stripe(secretKey, {
+      appInfo: {
+        name: 'stripe-mcp',
```
Comment (matv-stripe) [inline tools/typescript/src/modelcontextprotocol/toolkit.ts]: name is important for metrics

File: tools/typescript/src/shared/mcp-client.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    const headers: Record<string, string> = {
+      Authorization: `Bearer ${config.secretKey}`,
+    };
+
+    if (config.context?.account) {
+      headers['Stripe-Account'] = config.context.account;
+    }
+
+    if (config.context?.customer) {
+      headers['X-Stripe-Customer'] = config.context.customer;
```
Comment (matv-stripe) [inline tools/typescript/src/shared/mcp-client.ts]: this won't do anything

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 1 above)
Comment (matv-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts:32]: https://github.com/stripe/ai/blob/main/tools/typescript/src/shared/invoices/listInvoices.ts#L12-L20 , some of these use the customer parameter to filter out results per a specific customer if you pass that in

File: tools/typescript/src/ai-sdk/toolkit.ts [hunk 4]
```diff
@@ -26,38 +26,48 @@ type StripeMiddlewareConfig = {
...
+  }
+
+  close(): Promise<void> {
+    return super.close({});
+  }
+
+  /**
+   * Middleware for billing based on token usage.
+   * Note: This uses direct Stripe SDK calls, not MCP.
+   */
```
Comment (matv-stripe) [inline tools/typescript/src/ai-sdk/toolkit.ts]: I wonder if we should delete this (not in this PR, just eventually), and point users to https://docs.stripe.com/billing/token-billing

File: tools/typescript/src/modelcontextprotocol/toolkit.ts [hunk 5]
```diff
@@ -1,58 +1,167 @@
...
+    this._configuration = configuration;
+
+    // MCP client for connecting to mcp.stripe.com
+    this._mcpClient = new StripeMcpClient({
+      secretKey,
+      context: {
+        account: configuration.context?.account,
+        customer: configuration.context?.customer,
+      },
+      mode: configuration.context?.mode,
```
Comment (matv-stripe) [inline tools/typescript/src/modelcontextprotocol/toolkit.ts:34]: todo for me during review - what is mode

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 2 above)
Comment (matv-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts]: let's keep it as stripe-agent-toolkit

File: tools/typescript/src/shared/constants.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+ * Shared constants for the Stripe Agent Toolkit.
+ * VERSION is injected at build time from package.json via tsup.config.ts
+ */
+
+declare const process: {env: {PACKAGE_VERSION?: string}};
+
+export const VERSION = process.env.PACKAGE_VERSION || '0.0.0-development';
+export const MCP_SERVER_URL = 'https://mcp.stripe.com';
+export const TOOLKIT_HEADER = 'stripe-agent-toolkit-typescript';
+export const MCP_HEADER = 'stripe-mcp';
```
Comment (matv-stripe) [inline tools/typescript/src/shared/constants.ts]: maybe `stripe-mcp-local` ?

Comment (matv-stripe) [reply tools/typescript/src/shared/constants.ts]: once we bump this toolkit we need to make sure we are not accidentally breaking the local mcp server

File: tools/typescript/src/shared/mcp-client.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+    if (!key.startsWith('sk_') && !key.startsWith('rk_')) {
+      throw new Error(
+        'Invalid API key format. Expected sk_* (secret key) or rk_* (restricted key).'
+      );
+    }
+
+    if (key.startsWith('sk_')) {
+      console.warn(
+        '[DEPRECATION WARNING] Using sk_* keys with agent-toolkit is deprecated. ' +
```
Comment (matv-stripe) [inline tools/typescript/src/shared/mcp-client.ts]: I think this error might show up in local MCP as well, since MCP uses agent-toolkit under the hood. maybe we can have a more generic name here?

File: tools/typescript/src/ai-sdk/toolkit.ts [hunk 8]
```diff
@@ -26,38 +26,48 @@ type StripeMiddlewareConfig = {
...
+    const tools: Record<string, ProviderTool> = {};
+
+    for (const remoteTool of mcpTools) {
+      const zodSchema = jsonSchemaToZod(remoteTool.inputSchema);
+
+      tools[remoteTool.name] = tool({
+        description: remoteTool.description || remoteTool.name,
+        inputSchema: zodSchema,
+        execute: (args: z.infer<typeof zodSchema>) => {
+          return this.stripe.run(remoteTool.name, args);
```
Comment (matv-stripe) [inline tools/typescript/src/ai-sdk/toolkit.ts]: what is `this.stripe` here?

File: tools/typescript/src/shared/configuration.ts [hunk 9]
```diff
@@ -67,3 +67,75 @@ export const isToolAllowed = (
...
+ * (via restricted API keys) are the primary security boundary.
+ *
+ * To maintain security:
+ * 1. Always use restricted API keys (rk_*) with minimal permissions
+ * 2. Update this map when new tools are added to mcp.stripe.com
+ * 3. Consider the actions configuration as a convenience filter, not a security boundary
+ */
+const toolPermissionMap: Record<
+  string,
+  Array<{resource: Object; permission: Permission}>
```
Comment (matv-stripe) [inline tools/typescript/src/shared/configuration.ts]: i wonder if we should deprecate this tool permission mapping? (or is needed for a migration?)

since restricted keys permission will filter the list of tools, I think maybe we push users to use that to limit their tools?

File: tools/typescript/src/ai-sdk/toolkit.ts (hunk 8 above)
Comment (matv-stripe) [reply tools/typescript/src/ai-sdk/toolkit.ts]: I see now, i think it is a little confusing between stripe being the MCP Client object vs the actual SDK client

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 5 above)
Comment (matv-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts:34]: alright yeah makes same

File: tools/typescript/src/cloudflare/index.ts [hunk 10]
```diff
@@ -21,12 +21,43 @@ export type PaymentProps = {
...
   State extends PaymentState,
   Props extends PaymentProps,
 > extends McpAgent<Bindings, State, Props> {
+  /**
+   * Register a paid tool that requires payment before execution.
+   *
+   * Note: This uses registerPaidTool which makes direct Stripe SDK calls
+   * for billing operations (customer lookup, checkout sessions, etc.).
+   */
   paidTool<Args extends ZodRawShape>(
```
Comment (matv-stripe) [inline tools/typescript/src/cloudflare/index.ts:30]: i think we can delete this experiment as well

File: tools/typescript/src/cloudflare/index.ts [hunk 11]
```diff
@@ -21,12 +21,43 @@ export type PaymentProps = {
...
+ * ```typescript
+ * class MyPaidAgent extends experimental_PaidMcpAgent<Env, State, Props> {
+ *   async init() {
+ *     this.paidTool(
+ *       'premium_feature',
+ *       'A premium feature that requires payment',
+ *       { input: z.string() },
+ *       async (args) => ({ content: [{ type: 'text', text: 'Result' }] }),
+ *       {
+ *         paymentReason: 'Access premium features',
```
Comment (matv-stripe) [inline tools/typescript/src/cloudflare/index.ts]: i think this file should be entirely unchanged, its unrelated to the agent toolkit

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 1 above)
Comment (sgr-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts:32]: we now take customer in here and at the tool level and pass it into the tools that take customer

we prefer the customer at tool level or fall back to this one. We kept this one for backwards compatibility

File: tools/typescript/src/shared/configuration.ts (hunk 9 above)
Comment (sgr-stripe) [reply tools/typescript/src/shared/configuration.ts]: I don't think we can drop this for now without it being a breaking change - the comment above calls it out but we could add a deprecation notice to `isToolAllowedByName` if the user is manually configuring actions

File: tools/typescript/src/cloudflare/index.ts (hunk 10 above)
Comment (sgr-stripe) [reply tools/typescript/src/cloudflare/index.ts:30]: I removed that comment that was added. Do you want something else removed for this file as part of this change?

File: tools/typescript/src/shared/constants.ts (hunk 6 above)
Comment (sgr-stripe) [reply tools/typescript/src/shared/constants.ts]: Let me know what you're thinking here. Is there something that relies on that value?

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 2 above)
Comment (sgr-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts]: this ended up being fully removed

Comment (matv-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts]: nice

File: tools/typescript/src/modelcontextprotocol/toolkit.ts (hunk 1 above)
Comment (matv-stripe) [reply tools/typescript/src/modelcontextprotocol/toolkit.ts:32]: this lgtm now - my original comment was old!

File: tools/typescript/src/shared/constants.ts (hunk 6 above)
Comment (matv-stripe) [reply tools/typescript/src/shared/constants.ts]: i think this is good now --- i am mostly noting we will likely need to update https://github.com/stripe/ai/blob/main/tools/modelcontextprotocol/src/index.ts to work with this new toolkit, not specific to this value

File: tools/typescript/src/ai-sdk/toolkit.ts (hunk 8 above)
Comment (sgr-stripe) [reply tools/typescript/src/ai-sdk/toolkit.ts]: updated to `mcpClient`

File: tools/typescript/src/shared/mcp-client.ts (hunk 7 above)
Comment (sgr-stripe) [reply tools/typescript/src/shared/mcp-client.ts]: good call, I went more generic and "is being deprecated" instead of "is deprecated"

File: tools/typescript/src/ai-sdk/toolkit.ts (hunk 4 above)
Comment (sgr-stripe) [reply tools/typescript/src/ai-sdk/toolkit.ts]: done!

File: tools/typescript/src/shared/constants.ts (hunk 6 above)
Comment (sgr-stripe) [reply tools/typescript/src/shared/constants.ts]: oh good call - I'll do that once this merges

Comment (sgr-stripe) [reply tools/typescript/src/shared/constants.ts]: Actually, that's probably the thing users most use, so could be a larger effect of removing billing and the secret key deprecation - might think on it a little more

File: tools/typescript/src/shared/constants.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Shared constants for the Stripe Agent Toolkit.
+ * VERSION is injected at build time from package.json via tsup.config.ts
```
Comment (matv-stripe) [inline tools/typescript/src/shared/constants.ts:3]: nice improvemtn

File: tools/typescript/src/test/shared/mcp-client.test.ts [hunk 13]
```diff
@@ -0,0 +1,225 @@
+import {StripeMcpClient} from '@/shared/mcp-client';
+
+// Mock the MCP SDK
```
Comment (matv-stripe) [inline tools/typescript/src/test/shared/mcp-client.test.ts:3]: maybe it would be better to mock at the HTTP layer? I think this is okay as is though

File: tools/typescript/src/shared/configuration.ts (hunk 9 above)
Comment (sgr-stripe) [reply tools/typescript/src/shared/configuration.ts]: papertrail: after speaking on slack, this is going to be pulled out and noted in the migration.md
